src/validation.cpp

```diff
@@ -3622,17 +3622,30 @@ bool BlockManager::LoadBlockIndex(
         vSortedByHeight.push_back(std::make_pair(pindex->nHeight, pindex));
     }
     sort(vSortedByHeight.begin(), vSortedByHeight.end());
+
+    // Find start of assumed-valid region.
+    int first_assumed_valid_height = std::numeric_limits<int>::max();
+    for (const auto& [height, block] : vSortedByHeight) {
+        if (block->IsAssumedValid()) {
+            first_assumed_valid_height = height;
```

Comment:
nit: Maybe add a `assert(IsSnapshotActive()` as we shouldn't have assumed-valid blocks in the absence of a snapshot loaded ?

Reply:
Part of @ryanofsky's overarching (and good) feedback a few months earlier was to avoid leaking knowledge of the idea of "UTXO snapshots" in places where it doesn't need to be. Instead we should rely on more fundamental state (like the presence of assumed-valid block index entries). I think referencing snapshots here would be going the other direction.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/23174/commits/b4919698ff3cc7183ae9ada7026d7300720f3fcc#r748949660

> nit: Maybe add a `assert(IsSnapshotActive()` as we shouldn't have assumed-valid blocks in the absence of a snapshot loaded ?

I guess like James said I'd be a little disinclined to add an assert checking for assumeutxo snapshots in validation code that's dealing more abstractly with chainstates. But I could also see the assert being useful as a sanity check, so no strong opinion either way.

Another option would be to write the assert a more generically with something like:

```c++

// assert that the assumed valid blocks come from some assumed-valid chainstate.
assert(any_of(chainstates, [](auto& chainstate) { return chainstate->reliesOnAssumedValid(); });

// assert that the assumed valid blocks will be validated by some not assumed-valid chainstate
assert(any_of(chainstates, [](auto& chainstate) { return !chainstate->reliesOnAssumedValid(); });
```

Reply:
Good idea! Added.